src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs

```diff
@@ -0,0 +1,553 @@
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.CodeAnalysis.Sarif.Writers
+{
+    public static class PrereleaseCompatibilityTransformer
+    {
+        // As the SARIF Technical Committee develops the SARIF specification, it
+        // releases incremental versions of the schema, with SemVer versions such as
+        // "2.0.0-csd.2.beta.2018-10-10". To avoid having to update the version strings
+        // in the test files every time this happens, we replace the version
+        // in the test file with the current version after reading the file
+        // into memory.
+        private const string VersionPropertyPattern = @"""version""\s*:\s*""[^""]+""";
+        private static readonly Regex s_VersionRegex = new Regex(VersionPropertyPattern, RegexOptions.Compiled);
+
+        private const string SchemaPropertyPattern = @"""\$schema""\s*:\s*""[^""]+""";
+        private static readonly Regex s_SchemaRegex = new Regex(SchemaPropertyPattern, RegexOptions.Compiled);
+
+        public static string UpdateToCurrentVersion(string prereleaseSarifLog, bool forceUpdate = false, Formatting formatting = Formatting.None)
+        {
+            bool modifiedLog = false;
+            JObject sarifLog = JObject.Parse(prereleaseSarifLog);
+
+            // Some tests update the semantic version to current for non-updated content. For this situation, we 
+            // allow the test code to force a transform, despite the fact that the provided version doesn't call for it.
+            string version = forceUpdate ? "2.0.0" : (string)sarifLog["version"];
+
+            switch (version)
+            {
+                // This case actually handles all SARIF v2 prerelease before we 
+                // started properly updating the schema. 
+                case "2.0.0":
+                {
+                    modifiedLog |= ApplyCoreTransformations(sarifLog);
+                    break;
+                }
+
+                case "2.0.0-csd.2.beta.2018-10-10":
+                {
+                    modifiedLog |= ApplyCoreTransformations(sarifLog);
+                    modifiedLog |= ApplyChangesSinceTechnicalCommitteeMeeting25();
+                    break;
+                }
+
+                default:
+                    break;
+            }
+
+            string text = modifiedLog ? sarifLog.ToString(formatting) : prereleaseSarifLog;
+
+            return text;
+        }
+        private static bool ApplyCoreTransformations(JObject sarifLog)
+        {
+            bool modifiedLog = UpdateSarifLogVersion(sarifLog); 
+
+            var runs = (JArray)sarifLog["runs"];
+
+            if (runs != null)
+            {
+                // Move result.ruleMessageId: https://github.com/oasis-tcs/sarif-spec/issues/216
+                // Update name of threadflowLocation.executionTimeUtc: https://github.com/oasis-tcs/sarif-spec/issues/242
+                modifiedLog |= UpdateRunResults(runs);
```

Comment:
>UpdateRunResults [](start = 31, length = 16)

Oh my heavens! I _love_ this approach!

Reply:
I organized this so that we can strongly associate TC changes with code, though you would approve. :)

---
In reply to: [226759993](https://github.com/Microsoft/sarif-sdk/pull/1114#discussion_r226759993) [](ancestors = 226759993)